### ISSUE TITLE
Please add a sensor expiry notification

### ISSUE THREAD

**Issue by Navid200** [NightscoutFoundation/xDrip#1467]
It would be great if there was the possibility of setting an alert for a given number of hours to the expiry time of a G6 sensor.
So, the user would be able to enter the hours to expiry as a parameter.
Even if it was possible to get an alert 12 hours to the expiry of the sensor, it would better than nothing.

**Comment by Navid200** [comment]
Reading my own description now, a few days later, I can see that the incentive may not be clear.

I get a notification, from xDrip, when the sensor stops.
But, if I get that notifications at 2AM, when I am sleep, it may not be the best time for restarting the sensor.

On the other hand, if I get a warning that the sensor will expire in 12 hours (at 2PM), that will give me a chance to plan my day with that in mind and plan to restart the sensor at the best possible time based on when I plan to eat and when I may have a meeting and when I may want to go to sleep.

**Comment by davidmtml** [comment]
I agree this would be a great thing.  Especially with the newer sensors where you can't easily restart, and can't fallback to xDrip algorithm.  I would like it to give a 24-hour heads up.  

"Your sensor will stop/expire in 24 hours".

This would give me a full day to figure out the best time to start a new sensor/restart the current one.

**Comment by tolot27** [comment]
Hence, a notification between 12 h to 24 h ahead of the expiry but during daytime?

**Comment by Navid200** [comment]
I would be careful about making it too complicated.  
If it is manageable to implement it such that the user has the choice of the warning time, that's perfect.  
If that's too difficult and it has to be decided ahead for everyone, I would just make it a constant period ahead of the expiry.

I would avoid "but during the day".  That would be discriminatory against individuals who work at night, like nurses, or firefighters etc.

**Comment by davidmtml** [comment]
User-defined would be great, but I think if it's going to be a set amount for everybody, 24 hours would be better than 12 for a few reasons.

1.  It just gives you a little bit more time to prepare, but not too much.
2.  This way, for most people, the alert would come at a time when you ARE already awake.  I work a normal schedule, so obviously I'm not putting in a new sensor at 4am or anything.  But I very well might put in a new sensor at 4pm.   I would prefer my notification to come 24 hours before, not 12 hours before at 4am.  I guess I would put it this way.  Take any time that you are awake.  Substract 12 hours.  There's a good chance you would be sleeping during that time.

**Comment by tolot27** [comment]
PR #1136 addresses the same feature request for Libre sensors (#570).

**Comment by Navid200** [comment]
@jamorham That's what we tell everyone they can use.
But, it would be great if there was an alarm option as well.

**Comment by Navid200** [comment]
Closing this as it is a nice-to-have and not a necessity.

**Comment by jamespaulley** [comment]
I would really like to see this implemented as I keep getting caught out by expired sensors

**Comment by tolot27** [comment]
> I would really like to see this implemented as I keep getting caught out by expired sensors

If you enable "Show Sensor Expiry" as described above, you will see the information all the time you are looking at your phone. An alarm would raise just once and there is no agreement when it should raise.

BTW: Contributions are welcome and you can implement it by yourself and submit a PR:

**Comment by Navid200** [comment]
@jamespaulley Would you please tell us exactly how the official app behaves?
One of the concerns of the developers is what happens if you are sleep when it is the time for the alert.
So, is it really an alert?  Or is it just a log, or what?

I am at a disadvantage that I have no experience with the official app.  Any detail you can provide with respect to how the alert on the official app behaves can be very helpful.

**Comment by jamespaulley** [comment]
Cool. The show sensor expiry option is nice but I find it easy to overlook and not take in. The official app gives a notification at 24 hours. It's not an alarm, just a standard Android notification. I think this is fine and how the feature should behave in xDrip. I think it should just say something like "Your sensor will expire in 24 hours". I think it should repeat once at 12 hours. This is what the Dex app does. It just serves as a reminder so you don't leave for work in the morning and get caught out by an expired sensor in the middle of the day. This happens to me a lot at the moment

**Comment by gabbo73** [comment]
Same to me, it seems really strange that a basic feature like this, implemented since ages even on the good old G4 receiver, has been missed here. why losing what we already had?
BTW, it just happened to me... sensor stopped while I'm at work, preemptive restart doesn't work anymore so I definitely would have been alerted 12 hours before (at least).

**Comment by Navid200** [comment]
It's not a good idea to have an alert for this.  Waking up someone to tell them that their sensor will expire in 12 hours can be very annoying.  

The user knows when the sensor life ends when he/she starts it.  xDrip has reminders.  Every Android phone has a calendar.
The user can enter a reminder using any of those.

**Comment by cyrusbyte** [comment]
Too bad this is still missing. Not like it can't be a silent alert or a android notification without a alarm.

**Comment by Navid200** [comment]
You can show sensor expiry on the main screen if you like.
https://navid200.github.io/xDrip/docs/Sensor-Expiry.html

My thinking is that if user can accomplish something already, in this case using a calendar, adding it to xDrip adds a complexity to the code that is not absolutely necessary.

**Comment by cyrusbyte** [comment]
Then what was the whole point of all of us asking for this feature? We all were well aware that it could be done another way but that is besides the point. The point is we wanted to simplify our lives by not having to. So it would still be nice to have this feature if by some miracle someone had free time to do it.

**Comment by gabbo73** [comment]
Totally agree. From a user experience point of view, we really need that
feature in the application itself. There is already an alert engine that
could be easily used for that purpose.


Il mar 10 gen 2023, 03:02 cyrusbyte ***@***.***> ha scritto:

**Comment by jonprogers** [comment]
> You can show sensor expiry on the main screen if you like. https://navid200.github.io/xDrip/docs/Sensor-Expiry.html
> 
Having used the 'Show Sensor Expiry' functionality since it was added I, personally, find this better than an alert as I can check sensor expiry at any time when scheduling activities. Having this information on the main graph screen is perfect for me as I look at this screen regularly. Yes, it took a while to build the habit, but I find that it was definitely worth the effort.

If anyone hasn't given this function a go I would recommend it...

**Comment by davidmtml** [comment]
I would also still like to see this implemented.  Showing sensor expiry on the main screen is nice, and definitely should not be gotten rid of!  However, I very rarely actually open xDrip on my phone.  I wear a FitBit with Glance watchface, so 90% of the time I'm just getting my numbers from there.  I also use the widget on my phone background, so almost all other times I get my reading from there.  There's just really no need for me to open up the whole app all that often. 

I certainly understand where you're coming from Navid, but I respectfully disagree with your take that if something is already able to be accomplished, we don't need to add anything additional.  If we can make the process easier or more automated, we should.  Sure you could add a calendar entry every time you start a new sensor, and set up some reminders of when it will expire.  But that is a manual process.  

As others have said, the alert engine is already in place, and the sensor expiry time is already a part of xDrip, so I would imagine it wouldn't be too terribly hard to bring the two together.

**Comment by gabbo73** [comment]
Totally agree with Davidmtml :)

Il mar 10 gen 2023, 21:38 davidmtml ***@***.***> ha scritto:

**Comment by jamorham** [comment]
I think an alert would be useful as a reminder that sensor change time is approaching. I can implement something like this.

**Comment by jamorham** [comment]
Apologies I mean a notification.

**Comment by Navid200** [comment]
@jamorham  And as soon as we add a notification, someone is going to say lets add a setting for it so that if someone doesn't like it, they can remove it.
And we end up with one more setting.

There are people who don't use xDrip because it has so many settings.  Will you allow me to remove old settings that are not needed any longer please?

**Comment by Navid200** [comment]
@jamorham I can add a notification.  Your time is needed on more critical items.  Can we talk?

**Comment by Navid200** [comment]
There is an existing feature that let's you see the date and time your sensor will expire on the main screen: https://navid200.github.io/xDrip/docs/Sensor-Expiry.html
When that is enabled, the feature works for Dexcom and Libre.

There is another existing feature request for Libre: https://github.com/NightscoutFoundation/xDrip/issues/570

When this feature is added, no produce a notification close to expiry, it can work for both Libre and Dexcom likewise.  Therefore, I am closing this feature request as the other one for Libre is older.

**Comment by Navid200** [comment]
xDrip now offers sensor expiry notifications as of February 11, 2023.
https://github.com/NightscoutFoundation/xDrip/releases/tag/2023.02.11

https://navid200.github.io/xDrip/docs/Alerts_page.html
